tests/unit/test_toolchain.cpp

```diff
@@ -93,12 +98,31 @@ void test_packed_enum()
     VerifyOrQuit(neighbor.mState == Thread::Neighbor::kStateValid, "Toolchain::OT_TOOL_PACKED failed 4\n");
 }
 
+void test_addr_size_cpp()
+{
+#ifdef X64
+    CompileTimeAssert(offsetof(otNetifAddress, mNext) == 40, "mNext should offset by 40 bytes from front");
+    CompileTimeAssert(sizeof(otNetifAddress) == 48, "otNetifAddress should be 48 (unpacked) bytes");
+#else
+    CompileTimeAssert(offsetof(otNetifAddress, mNext) == 36, "mNext should offset by 36 bytes from front");
```

Comment:
These magic numbers create a bit of a maintenance headache.  If someone changes the data structure in a valid way they then may need to change these hard coded values.  Furthermore, they must know how to change the values to create a successful result. 

What I think you really want is a way to compare the result of sizeof(..) and offset(...) from the cpp file to the same result in the c file below.  For that you probably can't get away with using CompileTimeAssert but instead must do something like;

VerifyOrQuit(sizeof(otNetifAddress) == GetCSizeofOtNetifAddress())

where the function GetCSizeofNetifAddress is implemented in test_toolchain_c.c